plugins/widget/plugin.js

```diff
@@ -3423,6 +3423,41 @@
 		}
 	}
 
+	// Detect situation when selection partially starts or ends in widget and remove such widget from further processing (#3498).
+	// Fix might be obsolete after (#3551).
+	function getSelectedHtmlWithFullWidgets( editor ) {
+		var selection = editor.getSelection( true ),
+			ranges = selection.getRanges(),
+			editable = editor.editable(),
+			range,
+			wrapper,
+			outputHtml = '';
+
+		for ( var i = 0; i < ranges.length; i++ ) {
```

Comment:
I was inspired by this method when I created this solution. https://github.com/ckeditor/ckeditor4/blob/t/3498/core/editor.js#L705

I think it's not required and in most cases, it will be a single range each time. I'll remove it, as you consider this for loop as increased complexity.